File: ios/Sasquatsh/ViewModels/DeckBuilderViewModel.swift
import SwiftUI

@Observable
@MainActor
final class DeckBuilderViewModel {
    // Deck metadata
    var deckName = ""
    var formatId: String? = nil
    var description = ""
    var powerLevel: Int? = nil
    var isPublic = false
    var commanderScryfallId: String? = nil

    // Card search
    var searchQuery = ""
    var searchResults: [ScryfallCard] = []
    var isSearching = false
    private var searchTask: Task<Void, Never>?

    // Deck cards
    var cards: [MtgDeckCard] = []

    // State
    var isLoading = false
    var isSaving = false
    var error: String?
    var isEditing = false

    private var deckId: String?
    private var services: ServiceContainer?

    func configure(services: ServiceContainer) {
        self.services = services
    }

    // MARK: - Card Search

    func searchCards(query: String) {
        searchTask?.cancel()
        guard query.count >= 2, let services else {
            searchResults = []
            return
        }
        searchTask = Task {
            try? await Task.sleep(for: .milliseconds(300))
            guard !Task.isCancelled else { return }
            isSearching = true
            do {
                searchResults = try await services.scryfall.searchCards(query: query)
            } catch {
                if !Task.isCancelled { searchResults = [] }
            }
            isSearching = false
        }
    }

    func clearSearch() {
        searchQuery = ""
        searchResults = []
        searchTask?.cancel()
    }

    // MARK: - Deck Management

    func addCard(_ card: ScryfallCard, board: String = "main") {
        if let index = cards.firstIndex(where: { $0.scryfallId == card.scryfallId && $0.board == board }) {
            // Increment quantity
            let existing = cards[index]
            cards[index] = MtgDeckCard(
                id: existing.id, deckId: existing.deckId,
                scryfallId: existing.scryfallId,
                quantity: existing.quantity + 1,
                board: existing.board, card: existing.card
            )
        } else {
            cards.append(MtgDeckCard(
                id: UUID().uuidString, deckId: deckId,
                scryfallId: card.scryfallId,
                quantity: 1, board: board, card: card
            ))
        }
    }

    func removeCard(at index: Int) {
        guard cards.indices.contains(index) else { return }
        cards.remove(at: index)
    }

    func updateQuantity(at index: Int, quantity: Int) {
        guard cards.indices.contains(index) else { return }
        if quantity <= 0 {
            cards.remove(at: index)
        } else {
            let card = cards[index]
            cards[index] = MtgDeckCard(
                id: card.id, deckId: card.deckId,
                scryfallId: card.scryfallId,
                quantity: quantity, board: card.board, card: card.card
            )
        }
    }

    // MARK: - Stats

    var totalCards: Int { cards.filter { $0.board == "main" }.reduce(0) { $0 + $1.quantity } }
    var creatureCount: Int { cards.filter { $0.board == "main" && $0.card?.isCreature == true }.reduce(0) { $0 + $1.quantity } }
    var landCount: Int { cards.filter { $0.board == "main" && $0.card?.isLand == true }.reduce(0) { $0 + $1.quantity } }
    var spellCount: Int { totalCards - creatureCount - landCount }

    var avgCMC: Double {
        let nonLands = cards.filter { $0.board == "main" && $0.card?.isLand != true }
        let total = nonLands.reduce(0.0) { $0 + (($1.card?.cmc ?? 0) * Double($1.quantity)) }
        let count = nonLands.reduce(0) { $0 + $1.quantity }
        return count > 0 ? total / Double(count) : 0
    }

    var mainDeckCards: [MtgDeckCard] { cards.filter { $0.board == "main" } }
    var sideboardCards: [MtgDeckCard] { cards.filter { $0.board == "sideboard" } }

    var cardsByType: [(type: String, cards: [MtgDeckCard])] {
        let types = ["Creatures", "Instants", "Sorceries", "Enchantments", "Artifacts", "Planeswalkers", "Lands", "Other"]
        return types.compactMap { type in
            let matching = mainDeckCards.filter { $0.card?.typeCategory == type }
            return matching.isEmpty ? nil : (type, matching)
        }
    }

    // MARK: - Load / Save

    func loadForEdit(deck: MtgDeck) {
        isEditing = true
        deckId = deck.id
        deckName = deck.name
        formatId = deck.formatId
        description = deck.description ?? ""
        powerLevel = deck.powerLevel
        isPublic = deck.isPublic ?? false
        commanderScryfallId = deck.commanderScryfallId
        cards = deck.cards ?? []
    }

    func saveDeck() async -> MtgDeck? {
        guard let services else { return nil }
        isSaving = true
        error = nil
        do {
            if isEditing, let deckId {
                let input = UpdateDeckInput(
                    name: deckName,
                    formatId: formatId,
                    description: description.isEmpty ? nil : description,
                    powerLevel: powerLevel,
                    isPublic: isPublic,
                    commanderScryfallId: commanderScryfallId
                )
                let deck = try await services.mtgDecks.updateDeck(id: deckId, input: input)
                isSaving = false
                return deck
            } else {
                let input = CreateDeckInput(
                    name: deckName,
                    formatId: formatId,
                    description: description.isEmpty ? nil : description,
                    powerLevel: powerLevel,
                    isPublic: isPublic
                )
                let deck = try await services.mtgDecks.createDeck(input: input)
                deckId = deck.id
                isEditing = true

                // Add cards
                for card in cards {
                    let cardInput = DeckCardInput(
                        scryfallId: card.scryfallId,
                        quantity: card.quantity,
                        board: card.board
                    )
                    try await services.mtgDecks.addCard(deckId: deck.id, card: cardInput)
                }

                isSaving = false
                return deck
            }
        } catch {
            self.error = error.localizedDescription
            isSaving = false
            return nil
        }
    }

    var isValid: Bool {
        !deckName.trimmingCharacters(in: .whitespaces).isEmpty
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/DeckBuilderViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.CreateDeckInput
import com.sasquatsh.app.models.DeckCardInput
import com.sasquatsh.app.models.MtgDeck
import com.sasquatsh.app.models.MtgDeckCard
import com.sasquatsh.app.models.ScryfallCard
import com.sasquatsh.app.models.UpdateDeckInput
import com.sasquatsh.app.services.MtgDeckService
import com.sasquatsh.app.services.ScryfallService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

data class DeckBuilderUiState(
    val deckName: String = "",
    val formatId: String? = null,
    val description: String = "",
    val powerLevel: Int? = null,
    val isPublic: Boolean = false,
    val commanderScryfallId: String? = null,
    val searchQuery: String = "",
    val searchResults: List<ScryfallCard> = emptyList(),
    val isSearching: Boolean = false,
    val cards: List<MtgDeckCard> = emptyList(),
    val isLoading: Boolean = false,
    val isSaving: Boolean = false,
    val error: String? = null,
    val isEditing: Boolean = false
) {
    val isValid: Boolean get() = deckName.trim().isNotEmpty()

    val totalCards: Int
        get() = cards.filter { it.board == "main" }.sumOf { it.quantity }

    val creatureCount: Int
        get() = cards.filter { it.board == "main" && it.card?.isCreature == true }.sumOf { it.quantity }

    val landCount: Int
        get() = cards.filter { it.board == "main" && it.card?.isLand == true }.sumOf { it.quantity }

    val spellCount: Int
        get() = totalCards - creatureCount - landCount

    val avgCMC: Double
        get() {
            val nonLands = cards.filter { it.board == "main" && it.card?.isLand != true }
            val total = nonLands.sumOf { (it.card?.cmc ?: 0.0) * it.quantity }
            val count = nonLands.sumOf { it.quantity }
            return if (count > 0) total / count else 0.0
        }

    val mainDeckCards: List<MtgDeckCard>
        get() = cards.filter { it.board == "main" }

    val sideboardCards: List<MtgDeckCard>
        get() = cards.filter { it.board == "sideboard" }

    val cardsByType: List<Pair<String, List<MtgDeckCard>>>
        get() {
            val types = listOf(
                "Creatures", "Instants", "Sorceries", "Enchantments",
                "Artifacts", "Planeswalkers", "Lands", "Other"
            )
            return types.mapNotNull { type ->
                val matching = mainDeckCards.filter { it.card?.typeCategory == type }
                if (matching.isNotEmpty()) type to matching else null
            }
        }
}

@HiltViewModel
class DeckBuilderViewModel @Inject constructor(
    private val mtgDeckService: MtgDeckService,
    private val scryfallService: ScryfallService
) : ViewModel() {

    private val _uiState = MutableStateFlow(DeckBuilderUiState())
    val uiState: StateFlow<DeckBuilderUiState> = _uiState.asStateFlow()

    private var deckId: String? = null
    private var searchJob: Job? = null

    fun updateDeckName(name: String) {
        _uiState.update { it.copy(deckName = name) }
    }

    fun updateFormatId(formatId: String?) {
        _uiState.update { it.copy(formatId = formatId) }
    }

    fun updateDescription(description: String) {
        _uiState.update { it.copy(description = description) }
    }

    fun updatePowerLevel(level: Int?) {
        _uiState.update { it.copy(powerLevel = level) }
    }

    fun updateIsPublic(isPublic: Boolean) {
        _uiState.update { it.copy(isPublic = isPublic) }
    }

    fun updateCommanderScryfallId(id: String?) {
        _uiState.update { it.copy(commanderScryfallId = id) }
    }

    fun updateSearchQuery(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
        searchCards(query)
    }

    // Card Search

    fun searchCards(query: String) {
        searchJob?.cancel()
        if (query.length < 2) {
            _uiState.update { it.copy(searchResults = emptyList()) }
            return
        }
        searchJob = viewModelScope.launch {
            delay(300)
            _uiState.update { it.copy(isSearching = true) }
            try {
                val results = scryfallService.searchCards(query)
                _uiState.update { it.copy(searchResults = results, isSearching = false) }
            } catch (_: Exception) {
                _uiState.update { it.copy(searchResults = emptyList(), isSearching = false) }
            }
        }
    }

    fun clearSearch() {
        searchJob?.cancel()
        _uiState.update { it.copy(searchQuery = "", searchResults = emptyList()) }
    }

    // Deck Management

    fun addCard(card: ScryfallCard, board: String = "main") {
        _uiState.update { state ->
            val existingIndex = state.cards.indexOfFirst {
                it.scryfallId == card.scryfallId && it.board == board
            }
            if (existingIndex >= 0) {
                val existing = state.cards[existingIndex]
                val updated = existing.copy(quantity = existing.quantity + 1)
                val newCards = state.cards.toMutableList().apply { set(existingIndex, updated) }
                state.copy(cards = newCards)
            } else {
                val newCard = MtgDeckCard(
                    id = UUID.randomUUID().toString(),
                    deckId = deckId,
                    scryfallId = card.scryfallId,
                    quantity = 1,
                    board = board,
                    card = card
                )
                state.copy(cards = state.cards + newCard)
            }
        }
    }

    fun removeCard(index: Int) {
        _uiState.update { state ->
            if (index in state.cards.indices) {
                state.copy(cards = state.cards.toMutableList().apply { removeAt(index) })
            } else state
        }
    }

    fun updateQuantity(index: Int, quantity: Int) {
        _uiState.update { state ->
            if (index !in state.cards.indices) return@update state
            if (quantity <= 0) {
                state.copy(cards = state.cards.toMutableList().apply { removeAt(index) })
            } else {
                val card = state.cards[index]
                val updated = card.copy(quantity = quantity)
                val newCards = state.cards.toMutableList().apply { set(index, updated) }
                state.copy(cards = newCards)
            }
        }
    }

    // Load / Save

    fun loadForEdit(deck: MtgDeck) {
        deckId = deck.id
        _uiState.update {
            it.copy(
                isEditing = true,
                deckName = deck.name,
                formatId = deck.formatId,
                description = deck.description.orEmpty(),
                powerLevel = deck.powerLevel,
                isPublic = deck.isPublic ?: false,
                commanderScryfallId = deck.commanderScryfallId,
                cards = deck.cards ?: emptyList()
            )
        }
    }

    fun saveDeck(onSuccess: (MtgDeck) -> Unit) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, error = null) }
            val state = _uiState.value

            try {
                if (state.isEditing && deckId != null) {
                    val input = UpdateDeckInput(
                        name = state.deckName,
                        formatId = state.formatId,
                        description = state.description.ifEmpty { null },
                        powerLevel = state.powerLevel,
                        isPublic = state.isPublic,
                        commanderScryfallId = state.commanderScryfallId
                    )
                    val deck = mtgDeckService.updateDeck(deckId!!, input)
                    _uiState.update { it.copy(isSaving = false) }
                    onSuccess(deck)
                } else {
                    val input = CreateDeckInput(
                        name = state.deckName,
                        formatId = state.formatId,
                        description = state.description.ifEmpty { null },
                        powerLevel = state.powerLevel,
                        isPublic = state.isPublic
                    )
                    val deck = mtgDeckService.createDeck(input)
                    deckId = deck.id
                    _uiState.update { it.copy(isEditing = true) }

                    // Add cards to the newly created deck
                    for (card in state.cards) {
                        val cardInput = DeckCardInput(
                            scryfallId = card.scryfallId,
                            quantity = card.quantity,
                            board = card.board
                        )
                        mtgDeckService.addCard(deck.id, cardInput)
                    }

                    _uiState.update { it.copy(isSaving = false) }
                    onSuccess(deck)
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage, isSaving = false) }
            }
        }
    }
}
